src/reducers/index.js

```diff
@@ -614,6 +667,202 @@ export const onTokenFetchHistoryRequested = (state, action) => {
+ */
+export const onProposalFetchSuccess = (state, action) => {
+  const { proposalId, data } = action;
+
+  const oldState = get(state.proposals, proposalId, { id: proposalId })
+
+  return {
+    ...state,
+    proposals: {
+      ...state.proposals,
+      [proposalId]: {
+        ...oldState,
+        status: PROPOSAL_DOWNLOAD_STATUS.READY,
+        updatedAt: new Date().getTime(),
+        data,
+      },
+    },
+  };
+};
+
+/**
+ * @param {String} action.proposalId - The proposalId to mark as failed
+ * @param {String} action.errorMessage - Error message
+ */
+export const onProposalFetchFailed = (state, action) => {
+  const { proposalId, errorMessage } = action;
+
+  const { password } = get(state.proposals, proposalId, { id: proposalId })
+
+  return {
+    ...state,
+    proposals: {
+      ...state.proposals,
+      [proposalId]: {
+        id: proposalId,
+        password,
+        status: PROPOSAL_DOWNLOAD_STATUS.FAILED,
+        errorMessage: errorMessage,
+        updatedAt: new Date().getTime(),
+      },
+    },
+  };
+};
+
+/**
+ * @param {String} action.tokenUid The token identifier to fetch
+ */
+export const onProposalTokenFetchRequested = (state, action) => {
+  const { tokenUid } = action
+
+  const oldState = get(state.tokensCache, tokenUid, {
+    tokenUid,
+    symbol: '',
+    name: '',
+    status: TOKEN_DOWNLOAD_STATUS.LOADING,
+  })
+
+  return {
+    ...state,
+    tokensCache: {
+      ...state.tokensCache,
+      [tokenUid]: {
+        ...oldState,
+        status: TOKEN_DOWNLOAD_STATUS.LOADING,
+        oldStatus: oldState.status,
+      }
+    }
+  }
+}
+
+/**
+ * @param {String} action.tokenUid - The tokenUid to mark as success
+ * @param {Object} action.data - The token information to store on redux
+ */
+export const onProposalTokenFetchSuccess = (state, action) => {
+  const { tokenUid, data } = action;
+
+  return {
+    ...state,
+    tokensCache: {
+      ...state.tokensCache,
+      [tokenUid]: {
+        tokenUid,
+        symbol: data.symbol,
+        name: data.name,
+        status: TOKEN_DOWNLOAD_STATUS.READY,
+      },
+    },
+  };
+};
+
+/**
+ * @param {String} action.tokenUid - The tokenUid to mark as failed
+ * @param {String} action.errorMessage - Error message
+ */
+export const onProposalTokenFetchFailed = (state, action) => {
+  const { tokenUid, errorMessage } = action;
+
+  return {
+    ...state,
+    tokensCache: {
+      ...state.tokensCache,
+      [tokenUid]: {
+        id: tokenUid,
+        symbol: '',
+        name: '',
+        status: TOKEN_DOWNLOAD_STATUS.FAILED,
+        errorMessage: errorMessage,
+      },
+    },
+  };
+};
+
+/**
+ * @param {String} action.proposalId - The new proposalId to store
+ * @param {String} action.password - The proposal password
+ * @param {Object} action.data - The proposal history information to store on redux
+ */
+export const onNewProposalGeneration = (state, action) => {
```

Comment:
On creating a new proposal, `onNewProposalGeneration` is used.
On importing an existing proposal, `onProposalAdded` was used.

Since this was confusing, refactored the latter to `onProposalImported`.